Modules/Logger/src/selxLoggerImpl.cxx

```diff
@@ -18,96 +18,155 @@
+  switch (level) {
+    case LogLevel::TRC:
+      return spdlog::level::level_enum::trace;
+    case LogLevel::DBG:
+      return spdlog::level::level_enum::debug;
+    case LogLevel::INF:
+      return spdlog::level::level_enum::info;
+    case LogLevel::WRN:
+      return spdlog::level::level_enum::warn;
+    case LogLevel::ERR:
+      return spdlog::level::level_enum::err;
+    case LogLevel::CRT:
+      return spdlog::level::level_enum::critical;
+    case LogLevel::OFF:
+      return spdlog::level::level_enum::off;
+    default:
+      itkGenericExceptionMacro( "Invalid log level." );
+  }
+}
 
+void
 LoggerImpl
-::~LoggerImpl()
+::SetLogLevel( const LogLevel& level ) {
+  spdlog::level::level_enum spdloglevel = this->ToSpdLogLevel( level );
+  spdlog::set_level( spdloglevel );
+  for( const auto& item : this->m_Loggers )
+  {
+    item.second->set_level( spdloglevel );
+  }
+}
+
+void
+LoggerImpl
+::SetPattern( const std::string& pattern )
+{
+  spdlog::set_pattern( pattern );
+}
+
+void
+LoggerImpl
+::SetSyncMode()
+{
+  spdlog::set_sync_mode();
+}
+
+void
+LoggerImpl
+::SetAsyncMode()
+{
+  spdlog::set_async_mode(this->m_AsyncQueueSize, this->m_AsyncQueueOverflowPolicy);
+}
+
+void
+LoggerImpl
+::SetAsyncQueueBlockOnOverflow(void)
 {
+  this->m_AsyncQueueOverflowPolicy = AsyncQueueOverflowPolicyType::block_retry;
 }
 
+void
+LoggerImpl
+::SetAsyncQueueDiscardOnOverflow(void)
+{
+  this->m_AsyncQueueOverflowPolicy = AsyncQueueOverflowPolicyType::discard_log_msg;
+}
 
 void
 LoggerImpl
-::AddConsole( Logger::FormatType format )
+::SetAsyncQueueSize( const size_t& queueSize )
 {
-  boost::log::add_console_log(
-    std::cout,
-    boost::log::keywords::format = format
-    );
-}
-
-
-// void
-// LoggerImpl::LoggerImpl
-// ::AddFile( FileNameType fileName, FormatType format )
-// {
-//   boost::log::add_file_log(
-//     boost::log::keywords::file_name = fileName,
-//     boost::log::keywords::format = format
-//   );
-// }
-
-// void
-// LoggerImpl::LoggerImpl
-// ::AddFile( FileNameType fileName, Logger::ChannelType channel, Logger::FormatType format )
-// {
-//   boost::log::add_file_log(
-//     boost::log::keywords::file_name = fileName,
-//     boost::log::keywords::filter = channel_filter == channel,
-//     boost::log::keywords::format = format
-//   );
-// }
+  this->m_AsyncQueueSize = queueSize;
+}
 
 void
 LoggerImpl
-::Log( SeverityType severity, Logger::MessageType message )
+::AsyncQueueFlush( void )
 {
-  boost::log::record record = this->m_Logger.open_record( boost::log::keywords::severity = severity );
-  if( record )
+  for( const auto& identifierAndLogger : this->m_Loggers )
   {
-    boost::log::record_ostream strm( record );
-    strm << message;
-    strm.flush();
-    this->m_Logger.push_record( boost::move( record ) );
+    identifierAndLogger.second->flush();
   }
-  else
+}
+
+void
+LoggerImpl
+::AddStream( const std::string& identifier, std::ostream& stream, const bool& forceFlush )
```

Comment:
You could consider to add the loglevel for this stream as argument too (with a default setting?). That would avoid the user having to call SetLogLevel() additionally.

Reply:
postponed to #112